### DISCUSSION TITLE
Berry and GPIO isolation (Feature request)

### DISCUSSION THREAD

**Discussion by Braehead** [arendst/Tasmota#19528] [Ideas]
My suggestion is to allow the isolation of GPIOs for use solely by Berry scripts, effectively devorcing these selected GPIO from the Tasmota core itself.

The idea being that template selected GPIOs (digital or analgue) be allocated for Berry use only.  In this state, these GPIOs will no longer be processed in anyway by the Tasmota core.  Thereby allowing these GPIOs to be fully controlled by the Berry scripts,  Currently access to GPIOs is far too slow for actual driver coding.  And often further processing by Tasmota causes issues.

Then the Tasmota class would need some new low level functions that can be used to drive these GPIOs directly.  Functions such a GPIO set/read/toggle, ADC read, DAC write, state change call back, etc.

This would allow a very large range of hardware drivers and direct hardware functions to be added via Berry and allow lower leverl customisations to be added to the Tasmota core at a user level.

Peoples thoughts?

**Comment by s-hadinger** [comment]
This is already the case for `Input` and `Interrupt`. They are just configured as inputs by Tasmota and left to Berry to manage.

Also, `Output Lo` and `Output Hi` are preconfigured as Low/High outputs and then just left to Berry to do whatever it wants.

One thing we could do is add numbers to Output Lo/Hi so you can reference them from Berry.

**Comment by sfromis** [comment]
Yes, configuring the pin as `Input` reserves it for Berry. Using the index, you can have, say, input number 3 being dedicated to a function, without the Berry code having to know the physical gpio number.

For the functions you ask for, see https://tasmota.github.io/docs/Berry/#gpio-module

**Comment by smhc** [comment]
I'm needing something similar for a limit switch on a DC motor. I need to spin the motor until I detect a change via a LM393 IR sensor and then cut the motor off. However Tasmota doesn't seem to have support for interrupts that would otherwise allow this to happen fast enough. (fast_loop is not fast enough either.. it's reading the pin while the motor is spinning that is the issue). I suspect something like shutters might also benefit from this.

I've cobbled something together using Berry that spins the motor a little bit, then does a digitalRead, spins a bit less a bit slower, reads, etc as it slowly inches towards the right position. This works but is pretty inefficient.

It would be great if there were something like gpio.attachInterrupt(self.xxx).

I'm also experimenting with adding a berry event callback for counters: https://github.com/arendst/Tasmota/commit/602ea10aed61420ba94c0c049b1286df9bfa5fa

But I think core Tasmota support for manipulating gpios at this level would be best. At the least a component type of "InputInterrupt" that generates this event for berry to consume.

**Reply by sfromis** [reply]
If you are not using newest Tasmota builds, the fast loop callback was a lot slower than now, and could be tweaked by setting the `Sleep` period. With current Tasmota, fast loop should be 200 times a second, enough for many things. But still not interrupt-driven.

**Reply by smhc** [reply]
I'm using the latest version with 5ms fast loops. But even if I do a hard loop (using while ...) it still misses the spot. I believe I had a similar issue with using bare-bones C code. Perhaps the digitalRead simply takes too long vs the callback? I didn't have this issue when using attach interrupt with 'CHANGE'. Perhaps there is implicit debouncing occurring, not sure. I can of course use a geared motor or just continue 'stepping' this one like I am.

Regardless, it would be really nice to have the interrupt functionality available to Berry. The main reason I want to use Tasmota+Berry is the ability to remotely maintain the code vs needing to reflash.

'analogRead' and 'tone' would also be really useful to have exposed.

**Reply by s-hadinger** [reply]
Sorry but you will never have access to interrupts from Berry. Interrupts have very high constraints that are not compatible with Berry. 

Honestly I don't understand what you are trying to achieve. C digitalRead is as fast as it can get since it's just a register read mapped to memory.

**Reply by smhc** [reply]
Ok fair enough, I did suspect this may be the case wrt interrupts and berry.
It's a bit of a digression; but for what it's worth I have a DC motor that spins a worm drive plus a plastic disc through an LM393 IR sensor and I need it to stop at a specific point to open and close a locking pin: https://i.imgur.com/5sLzexA.jpg

Using a simple switch and rules worked ok:
```
Rule1
On Event#close do Backlog SetOption60 1; Sleep 0; SwitchMode1 2; Power1 1 ENDON
On Event#open do Backlog SetOption60 1; Sleep 0; SwitchMode1 1; Power1 1 ENDON
On Power1#State=0 do Backlog SetOption60 0; Sleep 220 ENDON
```

But I have been trying to make it more accurate and reliable. I might repeat my tests with vanilla C just to confirm whether the interrupt "CHANGE" approach is actually different to a digitalRead in a hard loop. It doesn't help that the motor has some inertia and capacitors.

Regardless of the correctness of my approach, it'd still be nice to get some lower level GPIO access so I don't want to derail the discussion with my project. I might experiment with the Tasmota sources to get the 'Counter' component to switch off the associated power as I see that uses interrupts. I also might try to expose "tone()" as that might be more reliable than gpio.set_pwm+delay.

**Reply by sfromis** [reply]
More generally, I'd say that Tasmota is not really suitable if you have millisecond-critical real-time code, as Tasmota instead is focused on having many things going on at the same time without preemptive multitasking. 50 ms ticks works fine for human-level speeds, but not for precise timing of moving parts.

As interrupt handlers are not feasible in Berry, closest I can see would be to use the Counter feature based on interrupts, and then checking if the counter value increased. Still not realistic for control purposes without extending Tasmota/Berry to provide direct read of counter values, as the method of `value=tasmota.cmd('Counter1',true)['Counter1']` is certainly not fast.

Actually, I've also seen others ask about reading counters from Berry, @s-hadinger .....

**Reply by s-hadinger** [reply]
Got it, I will add support for Counters in Berry

**Reply by barbudor** [reply]
While it is not possible to run berry code within the interrupt context, it could make sense to have some king of signalization from the interrupt to the loop and to asynchornously trigger a callback in the loop context.
That could be usefull to catch a short event that could be missed in the loop polling.
Of course it doesn't insure any kind of low latency as Tasmota could be stuck somewhere else

**Reply by sfromis** [reply]
Also, the potential of Tasmota being busy with relatively slow things means that there could be several interrupts before the main loop runs again. The strategy of using a counter would allow Berry code (with the Berry Counter support) to deal with this.

The cost of having the fast loop callback check this should not be too bad, and still faster than the main loop ticks. Effectively, the updated counter acts as the signal that an interrupt happened.

**Reply by s-hadinger** [reply]
Done (but not tested yet): https://github.com/arendst/Tasmota/pull/19558

**Reply by smhc** [reply]
Cool, thanks. I attempted to try it out but must have loaded the wrong firmware somehow. Although it doesn't matter too much as having looked into the counter stuff some more I realised I can't use this for my project anyway. I need to dynamically switch between detecting the first falling edge and first rising edge (i.e "CounterMode 2"), which the counter component doesn't seem to support (https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xsns_sensor/xsns_01_counter.ino#L74). I imagine most counter use doesn't have this requirement however and this feature will still be useful.

I suppose my current method of stepping with a digitalRead is the best method with Tasmota.

**Reply by sfromis** [reply]
Maybe it could be useful for other purposes to have inverted counter pins counting rising edge instead of falling edge. With that it would be possible to connect the input to two counter pins with diodes, one counting falling edge, and the other counting rising edge.

**Reply by smhc** [reply]
I do think it would be useful for other scenarios. It'd effectively be the ability to use interrupt logging for the various CHANGE/RISING/FALLING flags, not necessarily just for counters / rotary encoders. Perhaps that's what the 'Interrupt' component should be doing? (It's a bit unclear to me what this component does any differently to 'Input').

**Reply by sfromis** [reply]
From what I can se, using `GPIO_INTERRUPT` pins is not "doing" anything, Comments in the source code says for Berry, but without any coding to have specific support:
`// Interrupt pins to be catched by Berry` 

When it comes to rotary encoders, Tasmota does have a specific driver to support those, including the ability to produce JSON payloads which can be handled by rules or Berry code.

More generally, millisecond-critical timing code is not a good fit for using Berry at all.

**Reply by smhc** [reply]
Yep fair enough. For what it's worth I've resolved my timing issues. I underestimated the purpose of the spring adding some load/resistance and I think the driver I was using didn't help either. I had similar overshoot problems when wiring the limit switch directly to the enable pin, bypassing the esp32, so the timing wasn't the issue.

I guess if you really needed a DC motor rigged up to a limit switch it might be better implemented as a driver, or just wiring it in directly. I was under the impression this is a relatively common mechanic and it might be something tasmota wants to natively support at some stage. But using berry works well enough.

**Reply by sfromis** [reply]
For simple cases of a limit switch, the "Tasmota way" would be to config it as a Switch/Button input, able to turn off a power channel, with ways to do it depending on exact expectations for behavior.
